### ISSUE TITLE
Stream.mapPipelinedJob has bottleneck in loop function

### ISSUE THREAD

**Issue by Szer** [Hopac/Hopac#185]
This line is a bottleneck:
https://github.com/Hopac/Hopac/blob/72de32046f073d870655dc9203419470dc73e307/Libs/Hopac/Stream.fs#L546

Why is happens? If some job would take longer than usual (for example hitting timeout on HTTP request), and `loop` function will get this "slow" reply channel from `mailbox`, it will hang for a long time and therefore stop looping. Meanwhile other jobs could finish successfully, but no one listens for their result, because `loop` is waiting on "slow" reply.

minimal repro:

```fsharp
open Hopac
open Hopac.Infixes

let jobMock i = //every 8th job takes 10x times more than others
    if i % 8 = 0 then 5000 else 500
    |> timeOutMillis
    >>-. i

#time
seq {1..100}
|> Stream.ofSeq
|> Stream.mapPipelinedJob 8 jobMock
|> Stream.iter
|> run
// it takes 16 seconds to finish
```

Here is fixed `mapPipelineJob`

```fsharp
let mapPipelinedJobFixed (degree: int) (x2yJ: 'x -> #Job<'y>) (xs: Stream<'x>) : Stream<'y> =
    if degree < 1 then
          failwithf "degree must be 1 or greater, given %d" degree
        elif degree = 1 then
          Stream.mapJob x2yJ xs
    else

    let closing, closed = IVar(), IVar()
    Stream.delay <| fun () ->
      let iCh, oCh = Ch(), Ch()

      let mutable usage = 0
      let rec loop() =
        Alt.choose [

          closed ^->. Stream.Nil

          oCh ^-> function
            | Value y -> usage <- usage - 1
                         Stream.Cons (y, loop ())
            | Exn e   -> raise e

          (if usage < degree then
            iCh ^=> fun x ->
              usage <- usage + 1
              Job.tryInDelay 
                (fun () -> x2yJ x)
                (Value >> Ch.send oCh)
                (Exn >> Ch.send oCh)
              |> Job.queue
              >>= loop
           else Alt.never())
          (if usage = 0 && IVar.Now.isFull closing 
              then Alt.always Stream.Nil
              else Alt.never())
        ] |> memo

      Job.tryIn (xs |> Stream.iterJob (Ch.give iCh))
             (fun () -> closing *<= () )
             (fun e  -> oCh *<- Exn e)
      |> Job.start >>= loop

#time
seq {1..100}
|> Stream.ofSeq
|> mapPipelinedJobFixed 8 jobMock
|> Stream.iter
|> run
// it takes 9 seconds to finish
```

I could prepare PR tomorrow
@haf , @forki

**Comment by haf** [comment]
Yes, let's PR so we can walk through the code, line by line.

Quick reading: why memoise the choose?

**Comment by Szer** [comment]
> Quick reading: why memoise the choose? It is never called outside deferred contexts and has no side-effects from strict evaluation.

Just because `loop` hast to return `Stream` which is memoized.
https://github.com/Hopac/Hopac/blob/72de32046f073d870655dc9203419470dc73e307/Libs/Hopac/Stream.fs#L23

This will be another PR because I have inhouse implementation of non-memoized streams based on channels. Streams are useful in working with data, but memoizaition just add memory burden.

**Comment by Szer** [comment]
`Streams` in `Hopac` are functional. After first iteration it memoizes results and won't fire any process in it again. This is understandable, but if you have reference on some tremendous sized Stream you will run out of memory pretty soon because of memoization. That's why Hopac need another non-persistent `Streams`. As I said its another topic

**Comment by Szer** [comment]
> Quick reading: why memoise the choose? It is never called outside deferred contexts and has no side-effects from strict evaluation.

Final note:
pretty much every single function in `Stream` module is memoized:

https://github.com/Hopac/Hopac/blob/72de32046f073d870655dc9203419470dc73e307/Libs/Hopac/Stream.fs#L129
https://github.com/Hopac/Hopac/blob/72de32046f073d870655dc9203419470dc73e307/Libs/Hopac/Stream.fs#L141

And a lot of `>>-*` and `>>=*` operators like here:
https://github.com/Hopac/Hopac/blob/72de32046f073d870655dc9203419470dc73e307/Libs/Hopac/Stream.fs#L254

**Comment by haf** [comment]
Yes, they are memoized but when those promises go out of scope they are GCed, right? So we'll only keep as many promises in memory as needed to give all the remaining history to the slowest stream reader.

**Comment by Szer** [comment]
> Yes, they are memoized but when those promises go out of scope they are GCed, right? So we'll only keep as many promises in memory as needed to give all the remaining history to the slowest stream reader.

Of course, but what if you don't need them at all in the first place?
In case you need to fetch 1M objects from datalake in parallel, map, groupBy and fold these groups, those Promises on heap won't add anything at all because you never reiterate this stream
